pkg/occlient/occlient.go

```diff
@@ -1572,46 +1478,169 @@ func (c *Client) GetOnePodFromSelector(selector string) (*corev1.Pod, error) {
 	return &pods.Items[0], nil
 }
 
-// RsyncPath copies local directory to directory in running Pod.
-func (c *Client) RsyncPath(localPath string, targetPodName string, targetPath string) (string, error) {
-	log.Debugf("Syncing %s to pod %s:%s", localPath, targetPodName, targetPath)
+// CopyFile copies single local file to the directory in running Pod.
+// if asFile is true, localFile will be file otherwise it is path to a directory
+// watchedFiles is list of changed files captured during `odo watch`
+func (c *Client) CopyFile(asFile bool, localFile string, targetPodName string, targetPath string, watchedFiles []string) error {
+	log.Debugf("Copying file %s to pod %s:%s", localFile, targetPodName, targetPath)
 
-	// TODO: do this without using 'oc' binary
-	args := []string{
-		"rsync",
-		localPath,
-		fmt.Sprintf("%s:%s", targetPodName, targetPath),
-		"--exclude", ".git",
-		"--no-perms",
-	}
+	dest := targetPath + "/" + path.Base(localFile)
+	reader, writer := io.Pipe()
+
+	go func() {
+		defer writer.Close()
+		err := makeTar(localFile, dest, writer, watchedFiles)
+		if err != nil {
+			os.Exit(-1)
+		}
+
+	}()
+
+	var cmdArr []string
 
-	output, err := c.runOcComamnd(&OcCommand{args: args})
+	if !asFile {
+		cmdArr = []string{"tar", "xf", "-", "-C", targetPath, "--strip", "1"}
+	} else {
+		cmdArr = []string{"tar", "xf", "-", "-C", targetPath}
+	}
+	err := c.ExecCMDInContainer(targetPodName, cmdArr, writer, writer, reader, false)
 	if err != nil {
-		return "", err
+		return err
+	}
+	return nil
+}
+
+// checkFileExist check if given file exists or not
+func checkFileExist(fileName string) bool {
+	_, err := os.Stat(fileName)
+	if os.IsNotExist(err) {
+		return false
 	}
+	return true
+}
+
+//makeTar function is copied from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp.go#L309
+func makeTar(srcPath, destPath string, writer io.Writer, watchedFiles []string) error {
```

Comment:
just `files` might be a better name than `watchedFile`.There should be also a comment explaining that if `files` is set than `srcPath` is ignored.